wallet/storage/src/internal/mod.rs

```diff
@@ -31,19 +39,133 @@ use wallet_types::{
     AccountTxId, KeychainUsageState, RootKeyContent, RootKeyId, WalletTx,
 };
 
+fn challenge_to_sym_key(
+    password: &String,
+    kdf_challenge: KdfChallenge,
+) -> Result<SymmetricKey, crate::Error> {
+    let KdfResult::Argon2id {
+        hashed_password, ..
+    } = hash_from_challenge(kdf_challenge, password.as_bytes())
+        .map_err(|_| crate::Error::WalletInvalidPassword())?;
+
+    let sym_key = SymmetricKey::from_raw_key(
+        SymmetricKeyKind::XChacha20Poly1305,
+        hashed_password.as_slice(),
+    )
+    .expect("must be correct size");
+
+    Ok(sym_key)
+}
+
+fn password_to_sym_key(password: &String) -> crate::Result<(SymmetricKey, KdfChallenge)> {
+    let mut rng = make_true_rng();
+    let config = KdfConfig::Argon2id {
+        // TODO: hardcoded values
+        config: Argon2Config::new(700, 16, 2),
+        hash_length: NonZeroUsize::new(32).expect("not 0"),
+        salt_length: NonZeroUsize::new(32).expect("not 0"),
+    };
+    let kdf_result = hash_password(&mut rng, config, password.as_bytes())
+        .map_err(|_| crate::Error::WalletInvalidPassword())?;
+    let KdfResult::Argon2id {
+        hashed_password, ..
+    } = &kdf_result;
+
+    let sym_key = SymmetricKey::from_raw_key(
+        SymmetricKeyKind::XChacha20Poly1305,
+        hashed_password.as_slice(),
+    )
+    .expect("must be correct size");
+
+    let challenge = kdf_result.into_challenge();
+
+    Ok((sym_key, challenge))
+}
+
 /// Store for wallet data, parametrized over the backend B
-pub struct Store<B: storage::Backend>(storage::Storage<B, Schema>);
+pub struct Store<B: storage::Backend> {
+    storage: storage::Storage<B, Schema>,
+    locked: bool,
+    encryption_key: Option<SymmetricKey>,
+}
 
 impl<B: storage::Backend> Store<B> {
     /// Create a new wallet storage
     pub fn new(backend: B) -> crate::Result<Self> {
-        let storage = Self(storage::Storage::new(backend).map_err(crate::Error::from)?);
+        let storage: storage::Storage<B, Schema> =
+            storage::Storage::new(backend).map_err(crate::Error::from)?;
+
+        let mut storage = Self {
+            storage,
+            locked: true,
+            encryption_key: None,
```

Comment:
Should these two fields be rolled into one somehow?
Possibly using an enum, very rough sketch:
```rust
enum KeyStorageState {
    Decrypted(SymmKey), // Keys are available and are decrypted (or decryptable?) using given key
    Plain,              // Keys are available in plaintext, no need to decrypt
    Locked,             // Keys locked, may or may not be encrypted
}
```

Or is `locked = true && encryption_key = Some(_)` an admissible state?

Reply:
created an EncryptionState with only Locked and Unlocked(Option<SymmKey>), removed Plain as it was making the code more verbose.